packages/@aws-cdk/aws-ssm/lib/parameter.ts

```diff
@@ -0,0 +1,220 @@
+import iam = require('@aws-cdk/aws-iam');
+import cdk = require('@aws-cdk/cdk');
+import ssm = require('./ssm.generated');
+
+/**
+ * An SSM Parameter reference.
+ */
+export interface IParameter extends cdk.IConstruct {
+  /**
+   * The ARN of the SSM Parameter resource.
+   */
+  readonly parameterArn: string;
+
+  /**
+   * The name of the SSM Parameter resource.
+   */
+  readonly parameterName: string;
+
+  /**
+   * The type of the SSM Parameter resource.
+   */
+  readonly parameterType: string;
+
+  /**
+   * Grants read (DescribeParameter, GetParameter, GetParameterHistory) permissions on the SSM Parameter.
+   *
+   * @param grantee the role to be granted read-only access to the parameter.
+   */
+  grantRead(grantee: iam.IRole): IParameter;
+
+  /**
+   * Grants write (PutParameter) permissions on the SSM Parameter.
+   *
+   * @param grantee the role to be granted write access to the parameter.
+   */
+  grantWrite(grantee: iam.IRole): IParameter;
+}
+
+/**
+ * A String SSM Parameter.
+ */
+export interface IStringParameter extends IParameter {
+  /**
+   * A reference to the value of the SSM Parameter resource.
+   */
+  readonly parameterValue: string;
+}
+
+/**
+ * A StringList SSM Parameter.
+ */
+export interface IStringListParameter extends IParameter {
+  /**
+   * A reference to the value of the SSM Parameter resource.
+   */
+  readonly parameterValue: string[];
+}
+
+/**
+ * Properties needed to create a new SSM Parameter.
+ */
+export interface ParameterProps {
+  /**
+   * The allowed pattern for values of this parameter.
+   *
+   * @default no validation is performed
+   */
+  allowedPattern?: string;
+
+  /**
+   * A user-friendly description for the parameter.
+   *
+   * @default none
+   */
+  description?: string;
+
+  /**
+   * Name of the parameter.
+   *
+   * @default a name will be generated by CloudFormation
+   */
+  name?: string;
+}
+
+/**
+ * Properties needed to create a String SSM parameter.
+ */
+export interface StringParameterProps extends ParameterProps {
+  /**
+   * The value of the parameter. It may not reference another parameter and ``{{}}`` cannot be used in the value.
+   */
+  value: string;
+}
+
+/**
+ * Properties needed to create a StringList SSM Parameter
+ */
+export interface StringListParameterProps extends ParameterProps {
+  /**
+   * The values of the parameter. It may not reference another parameter and ``{{}}`` cannot be used in the value.
+   */
+  value: string[];
+}
+
+/**
+ * Basic features shared across all types of SSM Parameters.
+ */
+export abstract class Parameter extends cdk.Construct implements IParameter {
```

Comment:
`ParameterBase` I think is the convention that Elad was going for.

Reply:
Yap indeed

Reply:
Alternatively, we can consider making `Parameter` concrete and exposing a less opinionated surface area (i.e. `value: string; type: enum`) and then add `StringParameter` and `StringListParameter` as convenience classes. This might be more idiomatic and conforms to our guidelines.

Reply:
What your proposing makes it impossible to use Parameter as the base type (value can't be both a `string` and `string[]`).